src/Database/LSMTree/Internal/RunReaders.hs

```diff
@@ -84,10 +84,12 @@ new ::
   -> IO (Maybe (Readers (FS.Handle h)))
 new fs hbio runs = do
     readers <- zipWithM (fromRun . ReaderNumber) [1..] runs
-    (readersHeap, firstReadCtx) <- Heap.newMutableHeap (catMaybes readers)
-    for firstReadCtx $ \readCtx -> do
-      readersNext <- newIORef readCtx
-      return Readers {..}
+    case catMaybes readers of
```

Comment:
and TBH, this doesn't look like an improvement. Instead of `for firstReadCtx` you have manually implemented `for (nonEmpty (catMaybes readers)) $ \xs -> ...

Reply:
I think it is beneficial to "push up" the emptiness checks into the API and require the user to witness non-emptiness. Over time, my hope is that we can push the emptiness checks "further and further up" until they reach "parse time" or "data creation." My thought pattern is along the lines of *Parse, Don't Validate* philosophy.

All that said, the manual emptiness check is a bit messy. I instead changed the code to utilize the `MaybeT` monad transformer and `iwither` from the `WitherableWithIndex` type-class. The `lsm-tree` project already transitively depends on the [`transformers`](https://hackage.haskell.org/package/transformers) and [`witherable`](https://hackage.haskell.org/package/witherable) packages via `aeson`.  I very much like how the code now reads. However the `lsm-tree` library com ponet would incur a slight dependency footprint increase from adding `witherable` as a direct dependency.

Reply:
I think that general principle has merit and the signatures or `newMutableHeap` and `newLoserTree` are nicer with `NonEmpty` in the sense that they are more precise about why creation might not return an element.

However, in this particular case pushing up the emptiness check would require us to assert that runs are non-empty. And even that doesn't help any more once we can initialise run readers an offset (lower bound key), where we just don't know if there will be matching entries in the run. That's still okay, I don't mind having to do a check before creating the heap/loser tree instead of afterwards, but I don't think the benefits here are huge.

And `witherable` etc. make the code less clear IMO, especially for someone not familiar with these concepts. The `for (nonEmpty (catMaybes readers))` Oleg suggested is a good alternative I think.

Reply:
I took both of your feedback into account and updated the PR to use
```
for (nonEmpty (catMaybes readers))
```
 as well as forgoing any calls to `witherable`.

Reply:
@phadej I do agree that the benefits are not really clear. Getting rid of one `[]` case in heap/tree creation, otherwise mainly moving things around. I still thing the new signature itself is better, so I'd take the change. Do you have objections to that, or specific things you'd like to be done differently? If not, could you update your review?

Reply:
I will merge the PR now to not let it get stale. Please tell me if you are unhappy with that.